Fix 443 dev

damhuonglan opened the pull request:
### What does this PR do?
fixing #443

### What have been modified?
+ Commands: `ph5tostationxml`, `ph5validate`
+ Checking for response file inconsistency following 2 formats: metadatatoph5 format `[das_model]_[sensor_model]_[sr][cha]` and resp_load format `[das_model]_[sr]_[srm]_[gain]`. Each part in the format will be checked, except for `[cha]` because different channel codes can use the same response file. Requirement for the check is character `_` has to be removed from model before added to the response file name in metadatatoph5 and resp_load command.
+ `ph5tostationxml`: 
       - In case of no response data can be returned because of error, not return empty response but throw error instead. Use flag `-E/--emp_resp` if want to add empty response for debugging
       - In case of other bugs, not create stationxml. Use flag `--stationxml_on_error` to create stationxml if bug present in the data
       - Multiprocessing has been removed  because when checking subfolders of the given path, `multiprocessing` make the logging messages show up randomly. Using `for loop` instead help users recognize which one the messages are for. That way `ph5tostationxml` can inform when never stationxml has or has not been created for a ph5 data.

### Checklist
- [ ] This PR is not directly related to an existing issue (which has no PR yet).
- [ ] All tests pass.
- [ ] Any new or changed features have are documented.
- [ ] Changes have been added to `CHANGELOG.txt` .
- [ ] First time contributors have added your name to `CONTRIBUTORS.txt` .

damhuonglan commented:
@timronan According to what you request to not return empty response, I fixed the old code so that it will:
  + Raise PH5toStationXMLError if  

    -  (1) All response file names are blank in response_t
    -  (2) A station’s response_table_n_i has no n_i for it in response_t
    -  (3) response_t has n_i=-1 (this n_i value is created by metadata when there is no response signal. I feel uncomfortable to raise error for this because the current ph5/test_data/ph5/master.ph5 has this value in response_t)
    -  (4) response_file_das_a is blank for the checked n_i
    -  (5) There is no signal for either response_file_das_a or response_file_sensor_a (if response_file_sensor_a isn’t blank)

  + Log the warning message and return the response data according to the response paths indicated in response_t if:
      - resonse_file_sensor_a is blank while model exists.
      - resonse_file_sensor/das_a is inconsistent with model(s), sample rate, sample_rate_multiplier, gain or char_code.

I also added flag -E/—emp_resp that allow to return empty response for the cases 2,3,4,5 for the debugging purpose.

It passes the test case that you created. Please let me know if you agree with that. Thank you.

timronan commented:
The stationxml response output by fix_443_dev, attached as fix_443_dev_1.xml, is significantly different than the stationxml response output by the rollback version, attached as status_quo.xml, of ph5toststionxml. The stationxml responses  output by fix_443_dev seems to be missing stages and only contain instrument sensitivity, this is not an accurate representation of a response and this pull request needs additional development. 

Can you explain why ph5tostationxml was changed by PR #383. Is it because the indexing is off between the response, array and das tables? If this is the case, which on e of these tables contains the correct index? Can you also explain how ph5tostationxml was working before PR #383? I want to understand the issue better before commenting on how to proceed. 

[fix_443_dev_xml.zip](https://github.com/PIC-IRIS/PH5/files/5732102/fix_443_dev_xml.zip)

timronan commented on ph5/clients/ph5tostationxml.py:
```diff
@@ -147,6 +147,10 @@ def get_args():
     parser.add_argument("--uri", action="store", default="",
                         type=str, metavar="uri")
+
+    parser.add_argument("-E", "--emp_resp", action='store', default=False,
```
Adding an --emp_resp option seems like a work around for a deeper level issue. Add this flag may be a good  solution because some ph5 experiments may actually have completely empty response tables but this seems like it needs further investigation. Would it ever make sense for a ph5 experiment to have instruments with no responses? In general all broadband instruments must have a response.

timronan commented on ph5/clients/tests/test_ph5tostationxml.py:
```diff
@@ -6,16 +6,18 @@
 import sys
```
These unit test seems quite a bit different. Does it seem safe to change a large portion of code and unit test. It's hard to tell but it seems like some refactoring has occurred. If this code was refactored, do we want to do this in the same PR that is changing logic as well?

damhuonglan replied:
All of the change in test_ph5tostationxml.py are because ph5tostationxml.py has been changed to not return empty response but raise error when there are errors for no response data or response table has no response filenames at all.
When the behaviors are changed from logging error to raising error the tests must be changed a lot.

timronan replied:
Thanks for this explanation.

timronan reviewed (changes requested):
The stationxml response output by fix_443_dev, attached as fix_443_dev_1.xml, is significantly different than the stationxml response output by the rollback version, attached as status_quo.xml, of ph5toststionxml. The stationxml responses  output by fix_443_dev seems to be missing stages and only contain instrument sensitivity, this is not an accurate representation of a response and this pull request needs additional development. 

Can you explain why ph5tostationxml was changed by PR #383. Is it because the indexing is off between the response, array and das tables? If this is the case, which on e of these tables contains the correct index? Can you also explain how ph5tostationxml was working before PR #383? I want to understand the issue better before commenting on how to proceed. 

[fix_443_dev_xml.zip](https://github.com/PIC-IRIS/PH5/files/5732102/fix_443_dev_xml.zip)

timronan commented on ph5/utilities/tests/test_ph5validate.py:
```diff
@@ -61,30 +61,29 @@ def test_check_array_t(self):
         self.assertEqual(resp_check_info, self.resp_check_info)
 
     def test_check_response_t(self):
-        with LogCapture() as log:
```
Was this check removed for a specific reason?

timronan commented on ph5/utilities/validation.py:
```diff
@@ -128,70 +134,48 @@ def check_response_info(info, ph5, unique_filenames_n_i, checked_data_files,
-    m_file = None
+    d_file = None
+    s_file = ''
     # check resp file from metadata
```
This logic seems to be different. Is there a way to test the new logical path and verify that it is working as expected?

timronan commented on ph5/utilities/validation.py:
```diff
@@ -128,70 +134,48 @@ def check_response_info(info, ph5, unique_filenames_n_i, checked_data_files,
-                return (m_path, None)
-            else:
-                return False
-
-    # check sensor
-    s_ret = check_resp_file_name(
-        Response_t, info, 'sensor', unique_filenames_n_i, errors, logger)
-    if s_ret is not None:
-        s_file, s_path = s_ret
-        if s_file not in checked_data_files.keys():
-            if check_resp_data(ph5.ph5, s_file, errors, logger):
-                checked_data_files[s_file] = True
-            else:
-                checked_data_files[s_file] = False
-                s_path = None
-        else:
-            if not checked_data_files[s_file]:
-                s_path = None
-    else:
-        s_file, s_path = None, None
-
-    # check das
-    ret = check_resp_file_name(
-        Response_t, info, 'das', unique_filenames_n_i, errors, logger, m_file)
-    if ret is not None:
-        d_file, d_path = ret
-        if d_file not in checked_data_files.keys():
-            if check_resp_data(ph5.ph5, d_file, errors, logger):
-                checked_data_files[d_file] = True
-                return (d_path, s_path)
-            else:
-                checked_data_files[d_file] = False
-                return (None, s_path)
-        else:
-            if checked_data_files[d_file]:
-                return (d_path, s_path)
-            else:
-                return (None, s_path)
-    else:
-        return (None, s_path)
```
This logic seems to be different. Is there a way to test the new logical path and verify that it is working as expected?

damhuonglan replied:
I had to fix this to meet your requirement of changes.
Before, response data were checked only when the file names in the response table were consistent with information in array table and the name return is the one created from the information in array table.
Now, any inconsistencies between response table and information in array table will be logged as warning and not interrupt the data checking. Then data will be checked and if it is passed, the response file names in the response table will be returned.

timronan replied:
As a clarification, are ph5_validate.log files still output if there are inconsistencies between response table and information in array table?

What does a warning look like? Can you provide an example?

timronan commented on ph5/utilities/tests/test_ph5validate.py:
```diff
@@ -61,30 +61,29 @@ def test_check_array_t(self):
         self.assertEqual(resp_check_info, self.resp_check_info)
 
     def test_check_response_t(self):
-        with LogCapture() as log:
-            log.setLevel(logging.ERROR)
-            ret = self.ph5validate.check_response_t(self.resp_check_info)
-            self.assertEqual(ret[0].error, set())
-            self.assertEqual(log.records, [])
-
```
Was this check removed for a specific reason?

timronan replied:
Updated and reincluded

damhuonglan commented:
@timronan I have run ph5tostationxml with fix_443_dev and still have <Stage/> information after <InstrumentSensitivity/>: 
[19-017.zip](https://github.com/PIC-IRIS/PH5/files/5737131/19-017.zip)
Can you pull again and run one more time with different name for xml please? I'm afraid the last time your run it has the error and didn't create xml and you open the xml you created before the last run.

timronan commented:
@damhuonglan if I use the flag --response than stages are output that are consistent, based off a diff, with the original PH5toStationXML program. Could you make the output stationxml level be response as the default behavior? By default, ph5tostationxml should output stationxml at a response level which includes all of the stages.

damhuonglan commented on ph5/utilities/tests/test_ph5validate.py:
```diff
@@ -61,30 +61,36 @@ def test_check_array_t(self):
     def test_check_response_t(self):
+        errors = {'array 004, station 0407, channel -2, response_table_n_i -1:'
+                  ' Metadata response with n_i=-1 has no response data.'}
         with LogCapture() as log:
-            log.setLevel(logging.ERROR)
+            log.setLevel(logging.WARNING)
             ret = self.ph5validate.check_response_t(self.resp_check_info)
-            self.assertEqual(ret[0].error, set())
+            self.assertEqual(set(ret[0].error), errors)
```
This I added error message for n_i=-1. This is the case when metadatatoph5.py catches that the response has no signal.

timronan replied:
It seems correct for the `n_i=-1` case to be and Error exception rather than a warning. If there is no response information in a PH5 experiment than that experiment should be flagged, and should not output stationxml metadata until the experiment is fixed. It is not acceptable to be serving response-less PH5 data. The SEED archive works similar to this. PH5 expirments must be complete for them to be available to the public. This logic will also improve the PH5 validation process.

damhuonglan replied:
In ph5tostationxml n_i=-1 will throw error if flag -E is not selected. However in ph5validate it should be logged as error because users want to see as most error as possible to fix it. If you want to change this to throw an error, we should discuss with DG.

timronan replied:
The exception handling should be different in ph5tostationxml and ph5validate. PH5toSationxml should be designed to output complete and correct metadata. It should not be designed as a testing tool. Ph5validate seems like it should be a data creation and testing tool so we would want it to run to completion and log errors rather than throwing program stopping exceptions.

damhuonglan replied:
It is that way now. I have flag -E for debugging. It also help me to write unittest for warning when I use ph5/test_data/ph5/master.ph5 for testing. That file has the case n_i=-1. If there is no way to bypass that case, I cannot test the warning for response file names.
Flag -E will not affect user work because they will not use that in normal case.

timronan replied:
It seems fine to have the flag -E case as a backdoor for testing.

damhuonglan commented on ph5/clients/tests/test_ph5tostationxml.py:
```diff
@@ -310,16 +384,13 @@ def tearDown(self):
-            self.assertEqual(log.records[0].msg,
-                             'All response file names are blank in response '
-                             'table. Check if resp_load has been run.')
+    def test_read_networks(self):
+        with self.assertRaises(ph5tostationxml.PH5toStationXMLError) as contxt:
+            self.parser.read_networks()
+            self.assertEqual(
+                contxt.exception.message,
+                'All response file names are blank in response '
+                'table. Check if resp_load has been run.')
```
I need your opinion on this. When all response filenames are blank in response table it may mean that resp_load hasn't been run, it may be for another cases. Let me know if you want to raise error here or just log it as error or warning.

timronan replied:
Is it possible to have a PH5 file that contains a response if respload has not been run?

damhuonglan replied:
resp_load is one step to create PH5. There is still the case with PH5 file with no resp_load run. However that file may not be submitted to DMC because it must be validated before that.

timronan replied:
If there is no response information in a ph5 experiment than an error exception should be thrown and ph5tostationxml should not output a stationxml file. Although this logic will force data creators to have fairly complete ph5 metadata before the stationxml version of the metadata can be created, this will help prevent incomplete expirments from slipping through the validation process and will flag any incomplete experiments that are currently stored in the IRIS data holdings. Deciding to use an error or warning for this case seems a trade off between the ease of data creation and data extraction. I would argue that an error should be used and the burden should be put onto the data creator, because they have the tools and knowledge to fix the data rather than a random user that may stumble upon this information in the future.

damhuonglan replied:
If you look at the test above, you can see that it is to test if the function raise error for that case (PH5toStationXMLErrorV) and the specific error message for that. There will be no stationxml file created at all.

timronan replied:
I think that this exception looks good. We could consider changing the exception's wording to something similar to:

> Response table does not contain any response file names. Check if resp_load has been run.

This is a lesser nuance that is not as important as the logic.

damhuonglan commented:
> Why is this flag being added? What is it's use case?

Our DG sometimes creates sub-directories under the main one for testing. They don't want to include files under testing folder when running ph5tostationxml. To avoid that they have to create a completely different directory instead of sub-directory. If flag -T is used, only file directly under paths listed in flag -p will be processed. If flag -T isn't used, everything still work like normal.
If you don't want that flag. I can revert that commit.

timronan commented:
We can leave this flag, but as stated earlier, we need to be targeted in our PRs. This flag seems extraneous to this specific PR and should be associated with it's own issue and PR. We need to be careful about trying to do bulk fixes in PH5 PRs . We need to be able to identify and roll back issues without change significant amounts of code. Also, it is easy for multiple logical changes to be lost in code review.

damhuonglan commented:
> we need to be targeted in our PRs.

So I will revert it and do this in another PR.

damhuonglan commented:
> I am slightly concerned about changing PH5toStationxml and Validation on the same branch and in the same pull request.

validation.py is a module that keeps the the common checks between ph5validate and ph5tostationxml. To change the response check in ph5tostationxml, I have to change the functions in validation which results in the changes in ph5validate as well. That is the reason I have to change 3 of them as the same time. 
If you don't want to change all 3 files at the same time, I will move those functions to ph5tostationxml and test those functions in ph5tostationxml.

timronan commented:
@damhuonglan point taken on validation.py. Thanks for the explanation, you should leave it how it is already set up. I will work on a reviewing code change. Thanks for working with me.

timronan commented on ph5/clients/ph5tostationxml.py:
```diff
@@ -296,12 +302,9 @@ def __init__(self, manager):
         validation.check_resp_unique_n_i(
             self.manager.ph5, self.unique_errors, None)
-        self.resp_load_already = validation.check_resp_load(
-            self.manager.ph5.Response_t, self.unique_errors, None)
```
Where has this been moved? It seems that validation.check_resp_load has been updated and replaced. What function replaced it?

damhuonglan replied:
Function check_resp_load() has been changed to check_response_filename() and has been moved to Lines 674-678 inside read_networks(). I think it's better not to check response in __init__()

damhuonglan replied:
Function check_resp_load() has been changed to check_response_filename() at https://github.com/PIC-IRIS/PH5/blob/ccf4017bd9eb8dd327a4bdf5d4dfe65802841288/ph5/clients/ph5tostationxml.py#L674-L678 in read

I think it's better not to check response error in __init__().

timronan replied:
Thanks for this clarification.

timronan commented on ph5/utilities/ph5validate.py:
```diff
@@ -331,30 +331,33 @@ def check_response_t(self, resp_check_info):
-                self.ph5.Response_t, errors, LOGGER):
-            header %= len(errors)
-            return [ValidationBlock(heading=header, error=errors)]
+        if validation.check_has_response_filename(
```
Are validation.check_has_response_filename and validation.check_resp_load equivalent?

damhuonglan replied:
Holly recommended that the error message should include metadatatoph5(), not just resp_load(). So the name check_resp_load didn't cover that and was misleading. So I changed name to check_has_response_filename to be more descriptive.

timronan replied:
This make sense. Thanks for pointing this out.

timronan reviewed (commented):
This pull request seems close to being mergeable into the master branch. After all of the comments are addressed we should consider beta testing this PR before merging into the Master branch to verify that there are not any looming bugs that we are missing in the code.

damhuonglan commented:
https://github.com/PIC-IRIS/PH5/pull/445#discussion_r556189560
Example for the warning from inconsistencies between response table and information in array table:
https://github.com/PIC-IRIS/PH5/blob/ccf4017bd9eb8dd327a4bdf5d4dfe65802841288/ph5/clients/tests/test_ph5tostationxml.py#L382-L385
https://github.com/PIC-IRIS/PH5/blob/ccf4017bd9eb8dd327a4bdf5d4dfe65802841288/ph5/clients/tests/test_ph5tostationxml.py#L386-L389
https://github.com/PIC-IRIS/PH5/blob/ccf4017bd9eb8dd327a4bdf5d4dfe65802841288/ph5/clients/tests/test_ph5tostationxml.py#L392-L394

timronan commented on ph5/clients/ph5tostationxml.py:
```diff
@@ -100,9 +100,11 @@ def get_args():
+                        help=("Specify level of detail using NETWORK, "
+                              "STATION, CHANNEL, or RESPONSE."
+                              " Default: RESPONSE"),
+                        choices=('NETWORK', 'STATION', 'CHANNEL', 'RESPONSE'),
```
Testing verifies that this has been resolved.

timronan reviewed (commented):
All of this looks good. I am going to test this branch in a beta system until 01/20/2021. If we don't run into any issues I will approve this PR.

timronan commented:
Could the Response_T Warnings and Errors that are output by ph5_validate be moved under the station and channels levels to which they belong. Right now ph5_validate outputs these warnings and error as a disorganized list under the Response_T header:
```
-=-=-=-=-=-=-=-=-
Response_t
0 error, 240 warning, 0 info
-=-=-=-=-=-=-=-=-
WARNING: array 001, station 13, channel 3, response_table_n_i 0: response_file_das_a 'ZLAND_3C' is inconsistence with sensor_model='' and das_model='ZLAND3C'; sr=250 srm=1 gain=18 'cha=DPZ'.
WARNING: array 002, station 33, channel 3, response_table_n_i 0: response_file_das_a 'ZLAND_3C' is inconsistence with sensor_model='' and das_model='ZLAND3C'; sr=250 srm=1 gain=18 'cha=DPZ'.
WARNING: array 002, station 56, channel 1, response_table_n_i 0: response_file_das_a 'ZLAND_3C' is inconsistence with sensor_model='' and das_model='ZLAND3C'; sr=250 srm=1 gain=18 'cha=DPN'.
WARNING: array 002, station 39, channel 2, response_table_n_i 0: response_file_das_a 'ZLAND_3C' is inconsistence with sensor_model='' and das_model='ZLAND3C'; sr=250 srm=1 gain=18 'cha=DPE'.
WARNING: array 001, station 26, channel 3, response_table_n_i 0: response_file_das_a 'ZLAND_3C' is inconsistence with sensor_model='' and das_model='ZLAND3C'; sr=250 srm=1 gain=18 'cha=DPZ'.
WARNING: array 001, station 10, channel 1, response_table_n_i 0: response_file_das_a 'ZLAND_3C' is inconsistence with sensor_model='' and das_model='ZLAND3C'; sr=250 srm=1 gain=18 'cha=DPN'.
WARNING: array 003, station 104, channel 1, response_table_n_i 0: response_file_das_a 'ZLAND_3C' is inconsistence with sensor_model='' and das_model='ZLAND3C'; sr=250 srm=1 gain=18 'cha=DPN'.
```

I am proposing that they should look similar too:

```
-=-=-=-=-=-=-=-=-
Station 1 Channel 2
0 error, 2 warning, 0 info
-=-=-=-=-=-=-=-=-
WARNING: No station description found.
 WARNING: array 001, station 1, channel 2, response_table_n_i 0: response_file_das_a 'ZLAND_3C' is inconsistence with sensor_model='' and das_model='ZLAND3C'; sr=250 srm=1 gain=18 'cha=DPE'.
WARNING: Data exists after pickup time: 64 seconds.
-=-=-=-=-=-=-=-=-

```

This organization will make it easier to read the output from ph5_validate.

damhuonglan commented:
> Could the Response_T Warnings and Errors that are output by ph5_validate be moved under the station and channels levels to which they belong.

Let me try.

damhuonglan commented:
@timronan I've removed the Response_T Warnings and Errors that are output by ph5_validate to be under the station and channels levels to which they belong. Please look at that to see if it meets your requirements.

timronan commented:
@damhuonglan thanks for putting in this effort. I will do a review first thing next week 01/19/2021.

timronan commented:
The warnings now seems to be included with the stations and channels to which they belong, but the warning messages are not exact, and they are a little bit confusing. Fore example: I receive the warning message:

> WARNING: response_file_das_a 'ZLAND_3C' is inconsistent with sensor_model='' and das_model='ZLAND3C'; sr=250 srm=1 gain=18 'cha=DPE'.

For all channels in experiment 18-020. After doing a little bit of digging I found that 
`ph5tokef --Response_t -n master.ph5 -p ./` returns `response_file_das_a=/Experiment_g/Responses_g/ZLAND_3C`
`ph5tokef --all_arrays   -n master.ph5 -p ./` returns `das/model_s=ZLAND_3C`
and  `ph5tokef --Sort_t -n master.ph5 -p ./` returns nothing. 

The this warning is accurately being triggered but it is not directly telling us the problem. If fact the current message is telling us incorrect information saying that das_model='ZLAND3C'.  It should say something like:

> Station = ' ' Chan = ' ' Sort_T:sensor_model='' is inconsistent with Response_t:response_file_das_a='ZLAND_3C' and Das_T:'das/model_s='ZLAND_3C'

There should be a different message triggered by if else statements for each all of the combinations on differences between the tables. If this was made into a function we could potentially re-use in some of the other issues. 

The output messages should take information directly from the metadata tables. It should not be hard coded because of potential differences as displayed above.

damhuonglan commented:
The response_file_das_a must be in either resp_load format (`[das model]_[sample rate]_[sample rate multiplier]_[gain]`) or metadatatoph5 format (`[das model]_[sensor model]_[sample rate][channel code]`. According to the warning message, there is only das model in the response_file_das_a, which is wrong format.

If das model ='ZLAND', sensor model/ response_file_sensor_a is allowed to be blank.

Normally, I only see the the inconsistencies  with different model, sample_rate, etc. I haven't seen the problem with wrong format like this. Should I add the format guideline in the warning message to prevent this situation?
For example

> WARNING: response_file_das_a 'ZLAND_3C' is inconsistent with sensor_model='' and das_model='ZLAND3C'; sr=250 srm=1 gain=18 'cha=DPE'. Please check with format [das_model]_[sr]_[srm]_[gain] or [das_model]_[sensor_model]_[sr][cha].

timronan commented:
The foreign key between the Response_t, Sorts_T and Array_T when it comes to DAS Models is   Sorts_T:response_table_n_i==Array_T:response_table_n_i==Response_t:n_i. 

Why does response_file_das_a need to be output as the resp_load format or the metadatatoph5 format? It seems like the n_i should be enough to properly identify the correct response. Am I missing something?

ascire-pic commented:
I don't think it actually matters what format the response_file_das_a field is in, as long as it is populated with something valid, whether that is in the metadatatoph5 or resp_load format. I could think of a few cases where, after loading data with metadatatoph5, the array tables could then be edited to change the das or sensor model. That shouldn't be a problem as long as the n_i is pointing to a response.

ascire-pic commented:
As an additional note, I think the experiment Tim is referencing is a bit older and therefore the DAS file name field is not as complete as the ones we started using later. Ideally, to identify a unique response, you would need at the minimum the instrument type (ex ZLAND_3C, RT130_CMG3T), the sample rate, the gain (particularly for the ZLand or SmartSolos) and maybe the channel name (Less important because in the majority of cases, all channels have the same response, just a different orientation).

So based on that, maybe we should add a warning like Lan mentioned to indicate that a user should check and make sure it is okay. If you only have one configuration/instrument, then just having ZLAND_3C as the das_file would be fine, but if you have multiple responses due to multiple configurations, it might be worth the flag to the user that they should check and make sure they actually loaded the correct response. With the less descriptive das_file_name field, there could be some uncertainty that a user loaded the correct response, so the flag might be useful.

I would just make it a warning. I don't think it's something that should cause writing out stationXML or anything to fail.

timronan commented:
I thought this was an internal consistency check between the response tables and the array table but it seems that the check is between a set list of das models and what is contained in the metadata. Is that correct?

damhuonglan commented:
I just simply checked if the response filename in response table match with the name may be created from resp_load or metadatatoph5 because I think those are the standard ways to create rows in the Response_t. The information I got to check was from array table.

However it seems not necessary. If we want to change that, we need to set up rules for what to check, and expect to see the changes in code and test files.  Also, please give me the warning messages you want for each check when it's failed.

According to what Allisa said, I suggest the following rules:
+ response file name must be unique. (This will be checked by looking at response table only before looking at array table)
+ response file name must contain the model in the corresponding n_i the array table. (Some minor changes are allowed for model: ',-=._ '  can be added or removed from the model)
+ If you want to check for sample rate, gain, channel code in the filename, please add more to the rules. When we have the final rules, I will add them to the description of this PR then change the code for that.

Thank you for your contribution.

timronan commented:
@damhuonglan I really like this approach and completely agree. We should have specific  targeted fixes that are talked about through logical rules. Each PR should only be associated with a few logical rules or be associated with one or two test at most. This will help us keep our PR's targeted and concise. I think that that both 

> 

- response file name must be unique. (This will be checked by looking at response table only before looking at array table)

- response file name must contain the model in the corresponding n_i the array table. (Some minor changes are allowed for model: ',-=._ ' can be added or removed from the model)
seem like the correct rules for this PR. These are also very testable. 

If we want to check 

> check for sample rate, gain, channel code in the filename 

we should make another different PR.

damhuonglan commented:
Can we discuss about this on the next PH5 meeting? Bruce wants us to talk more about this. He'd like to have some input too.

hrotman-pic commented:
I have a question about the proposed sensor model consistency check.  This is from fields populated using the metadatatoph5 format.  In this example, the metadata is a stationXML created using Nexus.  These fields have not been modified after running metadatatoph5.

Field in array table:
sensor/model_s=L-28LB, 4.5 Hz, 27.0 V/m/s, Rc=395 Ohms, Rs=2490 Ohms

Field in response table:
response_file_das_a=/Experiment_g/Responses_g/RT130_L28LB45Hz270VmsRc395OhmsRs2490Ohms_500DH2

Will this fail the check?  If yes, please let me know if I should open an issue for it, or if we should wait until a PH5 meeting to discuss if it would be better to prevent it failing the check in the code or a different way.  Also please let me know if you need more information.

damhuonglan commented:
@hrotman-pic The following characters will be removed from sensor model before checking with response_file_das_a: ',' '-' '=' '.' '_' ' '. So the sensor model will turn into: L28LB45Hz270V/m/sRc395OhmsRs2490Ohms. 
It has a little bit problem here is the '/' is not removed. But I can fix it.
I think 500 is sample rate and DH2 is channel code. Correct?
Thanks for giving this so I can fix my code. I will need you to check this after the fix.

timronan commented:
> response_file_das_a=/Experiment_g/Responses_g/RT130_L28LB45Hz270VmsRc395OhmsRs2490Ohms_500DH2

Should not be an acceptable. If a program is automatically creating response_file_das_a in this way, then the program needs to be modified. The need  response_file_das_a field needs to be stripped all additional strings (units) of the special characters. 

The check needs to be very specific:

respload format: [das model]_[sample rate]_[sample rate multiplier]_[gain]
[string]_[int]_[int]_[int]
metadatatoph5 format: [das model]_[sensor model]_[sample rate][channel code]
[string]_[string]_[int][string]

If we start following a pattern of semi random strings there is no point for this validation check. 

Realistically these formats should be consistent but that is a different PR.

damhuonglan commented:
> Should not be an acceptable. If a program is automatically creating response_file_das_a in this way, then the program needs to be modified. The need response_file_das_a field needs to be stripped all additional strings (units) of the special characters.

@timronan If you look at the field in array table `sensor/model_s=L-28LB, 4.5 Hz, 27.0 V/m/s, Rc=395 Ohms, Rs=2490 Ohms` that @hrotman-pic quote, the `response_file_das_a=/Experiment_g/Responses_g/RT130_L28LB45Hz270VmsRc395OhmsRs2490Ohms_500DH2` is following the format `metadatatoph5 format: [das model]_[sensor model]_[sample rate]_[channel code]` if the following characters are omitted: `','  '-'  '='  '.'  '_'  ' '  '/'`.

So it should be acceptable if I add '/' in the list of characters to be ignored because others already in the list.

timronan commented:
Alright, I get it. As @damhuonglan suggest, all of the special characters should be stripped from the check and this should work. I don't like it that:

>  Field in array table:
sensor/model_s=L-28LB, 4.5 Hz, 27.0 V/m/s, Rc=395 Ohms, Rs=2490 Ohms

Is acceptable in the array table but I don't think there is anything we can do about that one.

hrotman-pic commented:
@damhuonglan thank you for looking at my example and explaining.  Do you want me to look at the other metadatatoph5 test archives I did a while ago and see if there are any additional special characters?

@timronan I don't like it being that way in the array table either.  For example, the array kef in question cannot be converted to a csv (keftocsv fails).  I have been considering opening an issue for metadatatoph5 to strip out commas during load (perhaps replaced with a field separator that won't cause problems?).  Or, open an issue to have a discussion about how to best prevent things we don't want in the array tables when we load metadata with metadatatoph5 without removing potentially needed/useful information about the das or sensor.  But, I'm not attached to being the one to start it, and there may be other problems with this example that I don't immediately see.

timronan commented:
@hrotman-pic you should start a new issue about metadatatoph5. It should say something about throwing an error if special characters or spaces are trying to be translated from metadata into ph5. It shouldn't be allowed. As you said, this causes bugs throughout our code stack that and this is not clean data. The user should have to figure out how to describe the field they are trying to describe without spaces or special characters.

damhuonglan commented:
@ascire-pic complained about the confusing error messages for blank sensor model when filename following metadatatoph5 format failed the check even for only one part of it. So I decided to:
+ split check_resp_file_name() into 3 different functions: check_metadatatoph5_format() check_das_resp_load_format() check_sensor() to be easier to process. 
+ In check_metadatatoph5_format() check_das_resp_load_format(), I split the filename into parts according to '_' and try to check each parts acording to metadatatoph5 or resp_load format. That way the error message will be a lot more concise and it can decise the format of the filename better to avoid extra confusing error message

For example with the case response_file_das_a='NoneQ330_NoneCMG3T_100LHN'; das_model='rt125a'


**The check in the old way will give 2 error messages:**

`
array 003 station 0407, channel 1: Response_t[6]:response_file_sensor_a is blank while sensor model exists.
`

`
array 003 station 0407, channel 1: Response_t[6]:response_file_das_a 'NoneQ330_NoneCMG3T_100LHN' is incomplete or inconsistent with Array_t_003:sensor_model=NoneCMG3T Array_t_003:das_model=rt125a Array_t_003:sr=100 Array_t_003:srm=1 Array_t_003:gain=1 Array_t_003:cha=DPZ. Please check with format [das_model]_[sr]_[srm]_[gain] or response_file_sensor_a 'gs11v' is inconsistent with Array_t_009:sensor_model=cmg3t.
`


**The check in the new way is only 1 message and more concise:**
`
array 003 station 0407, channel 1: Response_t[6]:response_file_das_a 'NoneQ330_NoneCMG3T_100LHN' is inconsistent with Array_t_003:das_model=rt125a. Please check with format [das_model]_[sensor_model]_[sr][cha].
`
+ For metadatatoph5 format, channel code check will be skipped because different channel code may use the same response file (according to @ascire-pic )
+ The reason I didn't use this approach before is because the model sections in the filename may contain `_`. So I will create a PR to strip off `_` from model before adding to response filename in metadatatoph5 and resp_load commands.

@ascire-pic  Please pull `fix_443_dev` again and check if it is the way you want.
@timronan  I know you don't want to change too much in this PR. If you think I only need to add special characters in the list and finish with this I will revert this change and ask @ascire-pic to create issue about the confusing messages and add this changes to the PR that fix that issue. But at first I would like to hear from you to see if it's worth to make these changes.

timronan commented:
@damhuonglan I agree with at @ascire-pic the error message is too confusing. There should be functions to check both of the formats for historical metadata. As @damhuonglan stated the metadatatoph5 check should not check the channel codes. As was stated by @ascire-pic multiple channels can point to the same response. 


For the channel code problem and other reasons, the metadatatoph5 function and  resp_load function need to generate consistent `response_file_das_a` file names as is outlined and discussed in issue #448.

ascire-pic commented:
I just ran ph5_validate and ph5tostationxml on the test archive I was using that added responses via metadatatoph5 and with the changes @damhuonglan made, the output makes more sense. The test archive did not throw any incorrect errors with respect to the das_file_name field and the stationXML file looks correct and complete. Originally ph5_validate and ph5tostationxml were complaining about the channel name in the das_file_name field which should not have been an issue. That error is no longer present.

timronan commented:
What do we think about making this check to only verify that  [das model][sensor model][sample rate] [String][String][Int] is consistent between the the check_resp_file_name() and the array_table. Although the Response_T[N]:response_file_das_a field will have to be parsed, these fields seem to be the most unique identifiers in the the respload and metadata2ph5 formats. This will set us up to work on issue #448. If the sensor model is missing from both fields than that field would be skipped.

timronan reviewed (changes requested):
This PR does not work as expected. An error is thrown when there is a mismatch between the response_table_n_I field and the corresponding array table fields,  but stationxml document with a truncated response is also output.  This can be proven by the test data set found at:

PH5/ph5/test_data/ph5/response_table_n_i

Run:
`ph5tostationxml -p ./ -n master.ph5 --level RESPONSE` or 
`ph5tostationxml` -p ./ -n master.ph5`

and inspect the output. 

Furthermore, the new error message that is output:

> ph5.clients.ph5tostationxml - ERROR: 002-0407-1 response_table_n_i 0: Response das file name should be 'NoneQ330_100_1_1' or 'NoneQ330_NoneCMG3T_100LHN' instead of 'NoneQ330_NoneCMG3T_200HHN'.

Is considerable less clear than the old message. It does not give us explicit instructions on which fields are being compared. An external user likely will not know what this error means. Also, I am fairly sure that this error message provides both the respload response_table_ni format and the metatoph5  response_table_ni format, but the message does not make this clear so I am unsure. Commit 71b1462dff7a0661e61586f257c0c57c0764cb3f has better messaging. Note the response_table_name _n_i format (Respload or Metadatatoph5) needs to be called out in the message as well.

damhuonglan commented:
> Furthermore, the new error message that is output:
> 
> > ph5.clients.ph5tostationxml - ERROR: 002-0407-1 response_table_n_i 0: Response das file name should be 'NoneQ330_100_1_1' or 'NoneQ330_NoneCMG3T_100LHN' instead of 'NoneQ330_NoneCMG3T_200HHN'.


@timronan I don't know how you checked out and updated the PR. This message looks like the very old version, before we fix anything on this. Please check it. The new error message must be something like:
`
array 002, station 0407, channel 1: Response_t[5]:response_file_das_a 'NoneQ330_CMG3T_200HHN' is inconsistent with Array_t_002:sensor_model=NoneCMG3T. Please check with format [das_model]_[sensor_model]_[sr][cha].
`

timronan commented:
Thanks for the message. I'm checking again.

timronan commented:
Okay, I am unsure what happened yesterday, anyways, the message looks a lot better but there are still some issues. 

1: The actual value in the Array_T:das/model_s is inconsistent with the value that is output by the message. The message should not transform the field's content, it is confusing. In experiment 18-0202 the message states Array_t_001:das_model=ZLAND3C but if the array table is queried Array_t_001:das_model=ZLAND_3C. This needs to be resolved. The

2: Are we checking for the channel code or not? The message below is unclear when it comes to this idea. 
ERROR: array 002 station 0407, channel 1: Response_t[0]:response_file_das_a 'NoneQ330_NoneCMG3T_200HHN' is inconsistent with Array_t_002:sr=100. Please check with format [das_model]_[sensor_model]_[sr][cha]

3: Stationxml should not be output if an Error message is thrown in ph5tostationxml. It seems like the point of this pull request is that we are unsure of which Response_Table entry is associated with which channel. 

4: The format's name need to be called out. The message needs to say to check either the Resp_Load or Metadata to ph5 format.

ascire-pic commented:
I thought we were going to output stationxml even if the error occurred to avoid accidentally breaking any archives that already exist at the DMC?

timronan commented:
In general, errors should stop a program's execution. For 3: in the list above, the metadata are stored separately from the PH5 data in the archive and if this error does not automatically output stationxml than the metadata in the archive should not be affected. To address the concern about outputting stationxml from historical potentially incorrect experiments, a flag, --stationxml_on_error or something similar, should be developed that outputs stationxml, following the original lph5tostationxml pattern, even if an error is thrown.

damhuonglan commented:
@timronan so you want flag `--stationxml_on_error` to log error when there are errors. Otherwise, there are nothing will be logged?

timronan commented:
@damhuonglan no. This error should always be thrown and logged regardless of flags. That is the point of this validation check, correct? 

Stationxml files should only be output from ph5tostationxml if 1: this bug is not present in the data. Or the flag `--stationxml_on_error` is included in the function call.

timronan commented:
This version is working really well and we are very close to the end of this PR, except we need to discuss what should be done about the test that @timronan just pushed, test_ph5tostationxml:test_Response_NI_DUPLICATION. 

The response table, found at PH5/ph5/test_data/response_table_n_i_dup/Response_Table_Duplication.kef in this case has duplicated response n_i values. The --stationxml-on-error flag does not work for this test case. The master branch of ph5 outputs stationxml documents with a response table configured in this way so the flag --stationxml-on-error  should also output stationxml from a PH5 expirment with this metadata configuration. It seems like taking a FIFO logical approach is the best way to handle this issue. The first response with a unque n_i value should be used.

damhuonglan commented:
>  The --stationxml-on-error flag does not work for this test case.

Let look at your test case for duplicated response_n_i:
https://github.com/PIC-IRIS/PH5/blob/b7dc1c32047ce6dca221c33a857275cf4a9013c6/ph5/clients/tests/test_ph5tostationxml.py#L585-L591

In line 591, you pass "--stationxml_on_error" as argument for getParser. However, in getParser() line 41, station_xml_on_error is always True:
https://github.com/PIC-IRIS/PH5/blob/b7dc1c32047ce6dca221c33a857275cf4a9013c6/ph5/clients/tests/test_ph5tostationxml.py#L21-L44

The argument you passed is at the position of minlat, so minlat="--stationxml_on_error" which creates the error:
```
  File "/Users/field/Documents/GIT/PH5/ph5/core/ph5utils.py", line 142, in is_rect_intersection
    minlat) > float(latitude):
ValueError: could not convert string to float: --stationxml_on_error
```

I wrote getParser() to test the data return in case `stationxml_on_error=True`.

The below lines is to test if the flag `stationxml_on_error` work or not:
https://github.com/PIC-IRIS/PH5/blob/b7dc1c32047ce6dca221c33a857275cf4a9013c6/ph5/clients/tests/test_ph5tostationxml.py#L123-L143

Then to fix your test case, just simply remove line 591

https://github.com/PIC-IRIS/PH5/blob/b7dc1c32047ce6dca221c33a857275cf4a9013c6/ph5/clients/tests/test_ph5tostationxml.py#L591

timronan commented on ph5/clients/tests/test_ph5tostationxml.py:
```diff
@@ -331,23 +445,41 @@ def setUp(self):
+            "Check if resp_load has been run or if metadatatoph5 input "
+            "contained response information.")
+
+    def test_read_networks_response_duplication(self):
```
This is the test that I was trying to create, I previously missed it, and I will delete the test instances that I pushed last night. 
The problem is that the master branch will output a stationxml files from a ph5 experiment with duplicated response n_i values in the responses table. It seems like this behavior should likely be mimicked if the flag --stationxml_on_error=True is included in the ph5tostationxml call. I previously suggested that we use a  FIFO logical approach and am open to other ideas. What do you think?

timronan commented on ph5/clients/tests/test_ph5tostationxml.py:
```diff
@@ -331,23 +445,41 @@ def setUp(self):
+            "Check if resp_load has been run or if metadatatoph5 input "
+            "contained response information.")
+
+    def test_read_networks_response_duplication(self):
+        # response_t_dup_n_i.kef has n_i= 1,3,6 duplicated
+        testargs = ['delete_table', '-n', 'master.ph5', '-R']
+        with patch.object(sys, 'argv', testargs):
+            nuke_table.main()
+
+        kefpath = os.path.join(self.home,
+                               'ph5/test_data/metadata/response_t_dup_n_i.kef')
+        testargs = ['keftoph5', '-n', 'master.ph5', '-k', kefpath]
+        with patch.object(sys, 'argv', testargs):
+            kef2ph5.main()
+
+        self.ph5sxml, self.mng, self.parser = getParser(
+            self.tmpdir, 'master.ph5', 'NETWORK')
+        with self.assertRaises(ph5tostationxml.PH5toStationXMLError) as contxt:
+            self.parser.read_networks()
+        self.assertEqual(
+            contxt.exception.message,
+            'Response_t n_i(s) duplicated: 1,3,6')
```
This is the test that I was trying to create, I previously missed it, and I will delete the test instances that I pushed last night.

The problem is that the master branch will output a stationxml files from a ph5 experiment with duplicated response n_i values in the responses table. It seems like this behavior should likely be mimicked if the flag --stationxml_on_error=True is included in the ph5tostationxml call. I previously suggested that we use a FIFO logical approach and am open to other ideas. @damhuonglan what do you think?

damhuonglan replied:
response's n_i duplication is checked before each row in array table is checked (Line 686) and error will be thrown if there is any duplication, the same as when there is no response_file_das_a in response_t (Line691).  In case the error is thrown, no stationxml is created at all.

https://github.com/PIC-IRIS/PH5/blob/b7dc1c32047ce6dca221c33a857275cf4a9013c6/ph5/clients/ph5tostationxml.py#L686-L694

I did that because when using response_n_i from array_t to get a row in response_t with the corresponding n_i, n_i must be unique to know which row to get. 

The flag `--stationxml_on_error` is only applied when the errors are logged, but not thrown.

timronan replied:
The uniqueness problem is an issue, and I understand that if we output stationxml files from ph5 experiments with non-unique n_i value than there is no way to determine if the stationxml document is accurate. The problem is that there are PH5 data in the archive, experiment 18-030, that have this non-unique n_i issue and I am worried that the metadata will get stuck inside these experiments if we don't make some backdoor to access them. I am not sure of the original logic used in ph5tostationxml but we do need some way to output the data that are stored in all of our current ph5 expirments.

If there is no response information in the response table than an error should be thrown and no stationxml should be output.

damhuonglan replied:
Can you post the response_t of the experiment 18-030 here please?

timronan replied:
> #
#	Fri Feb 19 22:22:02 2021	ph5 version: 4.1.2
#
#   Table row 1
/Experiment_g/Responses_g/Response_t
	n_i=0
	bit_weight/value_d=7.45058059692e-05
	bit_weight/units_s=mV/count
	gain/units_s=dB
	gain/value_i=12
	response_file_a=
	response_file_das_a=/Experiment_g/Responses_g/ZLAND3C_1000_1_12
	response_file_sensor_a=

#   Table row 2
/Experiment_g/Responses_g/Response_t
	n_i=1
	bit_weight/value_d=0.16985
	bit_weight/units_s=V/count
	gain/units_s=dB
	gain/value_i=24
	response_file_a=
	response_file_das_a=/Experiment_g/Responses_g/RESPGEOD
	response_file_sensor_a=/Experiment_g/Responses_g/RESPGS11D

#   Table row 3
/Experiment_g/Responses_g/Response_t
	n_i=0
	bit_weight/value_d=7.45058059692e-05
	bit_weight/units_s=mV/count
	gain/units_s=dB
	gain/value_i=12
	response_file_a=
	response_file_das_a=
	response_file_sensor_a=

#   Table row 4
/Experiment_g/Responses_g/Response_t
	n_i=1
	bit_weight/value_d=0.16985
	bit_weight/units_s=V/count
	gain/units_s=dB
	gain/value_i=24
	response_file_a=
	response_file_das_a=
	response_file_sensor_a=

timronan replied:
It is possible there could be other examples of this in the archive... I bet most of them could be resolved though. The current logic could be the correct answer but I just want to make sure we talk about it before moving forward.

hrotman-pic replied:
> It is possible there could be other examples of this in the archive...

16-012 also has a duplicated n_i 

#
#       Thu Dec 17 20:08:42 2020        ph5 version: 4.1.2
#
#   Table row 1
/Experiment_g/Responses_g/Response_t
        n_i=0
        bit_weight/value_d=5.96046447754e-08
        bit_weight/units_s=volts/count
        gain/units_s=
        gain/value_i=32
        response_file_a=
        response_file_das_a=/Experiment_g/Responses_g/rt125_1000_1_32
        response_file_sensor_a=/Experiment_g/Responses_g/GS11V
#   Table row 2
/Experiment_g/Responses_g/Response_t
        n_i=0
        bit_weight/value_d=5.96046447754e-08
        bit_weight/units_s=volts/count
        gain/units_s=
        gain/value_i=32
        response_file_a=
        response_file_das_a=/Experiment_g/Responses_g/rt125_1000_1_32
        response_file_sensor_a=/Experiment_g/Responses_g/L28LB

damhuonglan replied:
The one that Tim gave won't make any problem with creating stationxml, because the duplicated ones that has no value for response_file_das_a and response_file_sensor_a are coming later
But the one that Holy gave will read the wrong response_file_sensor_a for L28LB because it always read row 1 when querying for n_i=0. Did you create this reponse_t before we fix resp_load?

hrotman-pic replied:
Yes, I originally made this archive in September and October 2019.  The array table for L28LB has response_table_n_i=1, so the xml will not output responses for L28LB at all, but if the array table was not correct the xml will be like you said.  Unfortunately I cannot remember how the mistake happened originally, but how to deal with it is probably more important.

damhuonglan replied:
I will modify ph5tostationxml to apply flag `--stationxml_on_error` for n_i duplication as Tim ask. In the log message, I will recommend user to run rerun resp_load to fix n_i for the ph5.

timronan replied:
I think that using the first response with a unique N_I value that appears in the response table is the correct way to handle this issue because we will at least know the specific response that is being output to the stationxml document.

damhuonglan commented on ph5/clients/tests/test_ph5tostationxml.py:
```diff
@@ -415,14 +562,58 @@ def test_create_obs_network(self):
+                                'ph5/test_data/ph5/response_table_n_i_dup')
+        self.ph5_path_eror = os.path.join(self.home,
+                                          datapath)
+        self.ph5sxml, self.mng, self.parser = getParser(self.ph5_path_eror,
+                                                        "master.ph5",
+                                                        "RESPONSE")
```
Flag --stationxml_on_error can be set by passing argument `stationxml_on_error=True/False`

damhuonglan commented on ph5/clients/tests/test_ph5tostationxml.py:
```diff
@@ -415,14 +562,58 @@ def test_create_obs_network(self):
+                                                        "master.ph5",
+                                                        "RESPONSE")
+        self.parser.add_ph5_stationids()
+        station = self.parser.read_stations()
```
Checking response_t's n_i duplication in read_network(), not in read_station(). (Please look at line code 476-498.

timronan commented:
When I run: `ph5tostationxml -n master.ph5 -p ./` both errors and the stationxml  are output:

```
[2021-02-22 22:35:51,845] - ph5.clients.ph5tostationxml - ERROR: 006-6046-1 response_table_n_i 0: response_file_sensor_a is blank while sensor model exists.
[2021-02-22 22:35:51,845] - ph5.clients.ph5tostationxml - ERROR: 006-6047-1 response_table_n_i 0: Response das file name should be 'Geode_1000_1_12' or 'Geode_GS11D_1000GPZ' instead of 'ZLAND3C_1000_1_12'.
[2021-02-22 22:35:51,845] - ph5.clients.ph5tostationxml - ERROR: 006-6047-1 response_table_n_i 0: response_file_sensor_a is blank while sensor model exists.
[2021-02-22 22:35:51,845] - ph5.clients.ph5tostationxml - ERROR: 006-6048-1 response_table_n_i 0: Response das file name should be 'Geode_1000_1_12' or 'Geode_GS11D_1000GPZ' instead of 'ZLAND3C_1000_1_12'.
[2021-02-22 22:35:51,845] - ph5.clients.ph5tostationxml - ERROR: 006-6048-1 response_table_n_i 0: response_file_sensor_a is blank while sensor model exists.
[2021-02-22 22:35:51,845] - ph5.clients.ph5tostationxml - ERROR: Response_t n_i(s) duplicated: 0,1
<?xml version='1.0' encoding='UTF-8'?>
<FDSNStationXML xmlns:iris="http://www.iris.edu/xml/station/1/" xmlns="http://www.fdsn.org/xml/station/1" schemaVersion="1.1">
  <Source>PIC-PH5</Source>
  <Sender>IRIS-PASSCAL-DMC-PH5</Sender>
  <Module>PH5 WEB SERVICE: metadata | version: 1</Module>
  <ModuleURI></ModuleURI>
  <Created>2021-02-22 22:35:52.137374</Created>
  <Network code="2E" endDate="2019-01-15T06:21:45.000000Z" startDate="2019-01-06T02:25:05.000000Z" iris:PH5ReportNum="18-030">
    <Description>Thwaites Margins</Description>
    <TotalNumberStations>173</TotalNumberStations>
    <Station code="1010" endDate="2019-01-15T02:44:29.000000Z" startDate="2019-01-06T03:41:12.000000Z">
      <Latitude unit="DEGREES">-79.421768</Latitude>
      <Longitude unit="DEGREES">-111.968945</Longitude>
      <Elevation unit="METERS">1770.2</Elevation>
      <Site>
        <Name>Converted from UTM Zone 12S</Name>
      </Site>
      <CreationDate>2019-01-06T03:41:12.000000Z</CreationDate>
      <TerminationDate>2019-01-15T02:44:29.000000Z</TerminationDate>
      <TotalNumberChannels>9</TotalNumberChannels>
      <SelectedNumberChannels>3</SelectedNumberChannels>
      <Channel code="GP1" endDate="2019-01-15T02:44:29.000000Z" locationCode="" startDate="2019-01-06T03:41:12.000000Z" iris:PH5ReceiverId="1010" iris:PH5Component="1" iris:PH5Array="001">
        <Latitude unit="DEGREES">-79.421768</Latitude>
        <Longitude unit="DEGREES">-111.968945</Longitude>
        <Elevation unit="METERS">1770.2</Elevation>
        <Depth unit="METERS">0.0</Depth>
        <Azimuth unit="DEGREES">0.0</Azimuth>
        <Dip unit="DEGREES">0.0</Dip>
        <SampleRate>1000.0</SampleRate>
        <Sensor>
```

Stationxml should only be output when stationxml_on_error is set to true.

damhuonglan commented:
@timronan The error messages go back to the old style again. Please run 'git log' and look for the last commit in this PR. I think your PH5 went back to the old version. Please look at this comment: https://github.com/PIC-IRIS/PH5/pull/445#issuecomment-778318308

timronan commented:
```
git pull
Already up-to-date.
(ph5) -bash-4.2$ 
(ph5) -bash-4.2$ git log
commit 9b8daaf4bbfad55cb17bfead1014670e6aa01a76
Merge: 39cf326 8fd0fee
Author: Lan <ldam@passcal.nmt.edu>
Date:   Mon Feb 22 11:44:20 2021 -0700
    merge master
```

I checked my result again with the same answer. 

```
ph5tostationxml -n master.ph5 -p ./
[2021-02-23 01:43:34,317] - ph5.clients.ph5tostationxml - ERROR: 004-4001-1 response_table_n_i 0: Response das file name should be 'Geode_1000_1_12' or 'Geode_GS11D_1000GPZ' instead of 'ZLAND3C_1000_1_12'.
[2021-02-23 01:43:34,318] - ph5.clients.ph5tostationxml - ERROR: 004-4001-1 response_table_n_i 0: response_file_sensor_a is blank while sensor model exists.
```

damhuonglan commented:
You should try from the folder of the code, run `python ph5tostationxml.py -n master -p path/to/ph5`.  Because if you look at test in test_ph5tostationxml:
https://github.com/PIC-IRIS/PH5/blob/9b8daaf4bbfad55cb17bfead1014670e6aa01a76/ph5/clients/tests/test_ph5tostationxml.py#L319-L326

What showing in the messages you show doesn't match with the test that passed. You have something wrong in git setting and the ph5 that is running.

timronan commented:
The version number changes, which is good, but I needed to run `pip install -e .` to use the new version. It is working now and I am beginning to test. Sorry for the confusion.

timronan reviewed (approved):
Everything seems to be working as expected. It seems like this branch is ready to be merged into master. @damhuonglan do you see any last clean up that needs to be done? I just removed two unused log files. I can merge when everything seems satisfactory. I think this PR should be squashed and merged so we can roll back if needed.

damhuonglan commented:
@timronan I have updated CHANGELOG. However, there is one of your test seem to not completed yet. That is test_Response_NI_DUPLICATION() in test_ph5tostationxml.py at line 597. I had some comments on that about testing for n_i duplication should be on read_network() instead of read_stations(). Please read that part and complete the test.
Thank you.

timronan commented:
@damhuonglan the test  `test_read_networks_response_duplication` is the coverage that I was trying to create in `test_Response_NI_DUPLICATION()`. I removed  `test_Response_NI_DUPLICATION()` so we don't double our testing coverage. Do you think we are ready to merge?

damhuonglan commented:
@timronan Yes. I have nothing against that. You can merge that if you feel good about that.

timronan commented:
After extensive planning, code review and testing we have deemed this branch to likely be working at expected, and decided it is time for it to be squash and merged.
